app/models/article.rb

```diff
@@ -964,9 +969,10 @@ def enrich_image_attributes
   end
 
   def remove_prohibited_unicode_characters
-    return unless title&.match?(PROHIBITED_UNICODE_CHARACTERS_REGEX)
+    return unless title&.match?(BIDI_CONTROL_CHARACTERS)
 
-    self.title = title.gsub(PROHIBITED_UNICODE_CHARACTERS_REGEX, "")
+    bidi_stripped = title.gsub(BIDI_CONTROL_CHARACTERS, "")
+    self.title = bidi_stripped if bidi_stripped.blank? # title only contains BIDI characters = blank title
```

Comment:
I like the new name for this regular expression! I think that maybe something isn't quite right here. The original version of this method removes any "prohibited" unicode character. This new method _only_ updates the article's title _if_ the `bidi_stripped` is blank — so, any title that has _some_ BIDI control characters would still have them after this method runs.

I'm not entirely sure what's gone wrong, but the accompanying "sanitizes the title" spec seems to be giving a false sense of security, checking it out in console with `pry`, the title does _not_ match `/u202b/` - before or after sanitizing:

```ruby
> title.match /\u202b/
=> nil
```

I _think_ the false-positive test is because `normalize_title` removes `/\u202b/`, so the test appears to be passing, even though the `remove_prohibited_unicode_characters` method does not seem to be working as intended.

If that's actually _intended_ here — maybe the idea is to update the behavior to only remove _some_ of the characters between `u202a` and `u202e` — then I'm not sure why the expanded regex is being used on line 972, where it would match a much _wider_ range of BIDI control characters than are being removed? 

At the least, the method name is now a bit misleading as to the underlying behavior.

Reply:
Thank you for your input and sorry for the confusion. Maybe you can think of a better method name than `remove_prohibited_unicode_characters` as it is now veering away from the intention of the implementer.

How I intend to make this method work is just filtering out titles that only contain BiDi controls. "BiDi controls" as such are not malicious or should be prohibited (I imply the #14909 solution was wrong); it saves text in some languages from otherwise looking gibberish. But there is no point of using them on their own, and it may result in bad things such as allowing people to create a post with apparently empty title (but as far as I see on the browser environment, the effect does not seem to spill out of the element, so there'd be no real "security" concerns).

BiDi controls that I intend to "prohibit" in my updated logic are only U+202A and U+202B, which are now disrecommended [by Unicode](https://www.unicode.org/reports/tr9/#Directional_Formatting_Characters). This is done by not including them to `TITLE_CHARACTERS_ALLOWED`. You may see in this comment https://github.com/forem/forem/issues/4969#issuecomment-937195057 that U+202C to U+202E are also deprecated, but this is misunderstanding of the context. The linked document only lists them as "not Suitable for use With Markup", where authors can replace them with tags (namely, [`<bdo>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdo)), which is not the case since we don't allow users to input HTML in the title. The handling of those codepoints may change over the time, so this is orthogonal to the function described in the previous paragraph.

Reply:
Okay, that makes a lot of sense, actually. Can we maybe start by adding test(s) that cover(s) the specific behavior(s) you'd like to see? I suspect that would be the best next steps, as then we are free to refactor & rename without worry that behavior gets lost.

Reply:
Do you have any idea what kind of additional tests are good to have? Currently there are two test cases in `article_spec.rb` and they assume respectively (after my update) that:

1. Remove U+202B but not U+202C (`"\u202bThis starts with BIDI embedding\u202c"` → `"This starts with BIDI embedding\u202c"`)
2. Reject BiDi controls-only title (`"\u202a\u202b\u202c\u202d\u202e"` → `Title can't be blank`)

Reply:
Hey @yheuhtozr , thank you for your pr :raised_hands: 
I think that most of the cases are covered in the tests. I would add a test including the characters that you have allowed, e.g.  `\u061c` is not in the tests (like "it's allowed", "it's not allowed as the only character").

Reply:
@lightalloy Included more BiDi marks in the tests as requested. Please check it out.

Reply:
@yheuhtozr Thanks for updating the test cases, that helps clarify this a lot! :+1: